sktime/utils/datetime.py

```diff
@@ -88,26 +89,54 @@ def _get_freq(x):
         return None
 
 
-def _shift(x, by=1):
+def _shift(x, by=1, return_index=False):
     """Shift time point `x` by a step (`by`) given frequency of `x`.
 
     Parameters
     ----------
     x : pd.Period, pd.Timestamp, int
         Time point
-    by : int
+    by : int, optional, default=1
+    return_index : bool, optional, default=False
+        whether to return an index element (False) or a pandas Index (True)
 
     Returns
     -------
     ret : pd.Period, pd.Timestamp, int
         Shifted time point
     """
-    assert isinstance(x, (pd.Period, pd.Timestamp, int, np.integer)), type(x)
-    assert isinstance(by, (int, np.integer)) or is_integer_index(by), type(by)
+    # deprecate in 0.13.0 and remove in 0.14.0, pd.Timestamp will not have freq
     if isinstance(x, pd.Timestamp):
+        warn("_shift no longer supports x: pd.Timestamp fom 0.14.0", DeprecationWarning)
         if not hasattr(x, "freq") or x.freq is None:
             raise ValueError("No `freq` information available")
         by *= x.freq
+        return x + by
+        # raise TypeError("_shift no longer supports x: pd.Timestamp")
+
+    # we ensure idx is pd.Index, x is first (and usually only) element
+    if isinstance(x, pd.Index):
+        idx = x
+        x = idx[0]
+    else:
+        idx = pd.Index([x])
+
+    # if we want index, we can simply use add dunder or shift
+    if return_index:
+        if idx.is_integer():
+            return idx + by
+        else:
+            return idx.shift(by)
+
+    assert isinstance(x, (pd.Period, pd.Timestamp, int, np.integer)), type(x)
+    assert isinstance(by, (int, np.integer)) or is_integer_index(by), type(by)
+
+    # if we want index element, we can still use add, except if timestamp
+    # deprecate this in 0.13.0 and remove in 0.14.0
+    if isinstance(x, pd.Timestamp):
```

Comment:
This block looks like a duplicate of lines 109-114... Is it intended?

Reply:
yes, that was a duplication - removed it

Reply:
yes, indeed. I looked into it again, and in fact it is *not* a duplication, it just looks like one.

Why: the initial `x` can be index, or index element.
The final `x` is always an index element.

Logic-wise, the cases the two pieces of code treat are different:
* the initial `x` was a `pd.Timestamp`. In this case, we will not have a `freq` after pandas deprecation, and a warning (later exception) should be raised. This is treated in the first paragraph, and will be deprecated once we move away from `freq` in `pd.Timestamp`.
* the initial `x` was a time-like index, and its element is `pd.Timestamp`. In this case, we have a `freq` right now, and after deprecation. This is treated in the second paragraph, and should not be deprecated.

So, the logic was actually not redundant, but the exception raised the second time was (it would never be reached), and the deprecation action was incorrect (second paragraph should not be deprecated).

Hope that makes sense?